core/frontend/src/views/NetworkTestView.vue

```diff
@@ -0,0 +1,226 @@
+      this.websocket.onmessage = (message: MessageEvent): void => {
+        this.latency_us = window.performance.now() - parseInt(message.data, 10)
+      }
+      this.websocket.onerror = () => {
+        this.latency_us = -1
+      }
+
+      this.interval = setInterval(() => {
+        if (this.websocket === undefined || this.websocket?.readyState === WebSocket.CLOSED) {
+          this.openWebSocket()
+        }
+
+        if (this.websocket?.readyState === WebSocket.OPEN) {
+          this.websocket?.send(window.performance.now().toString())
+        }
+      }, 200)
+    },
+    start(): void {
+      this.updateState(State.DownloadSpeed)
+    },
+    clearSpeed(): void {
+      this.speed = 0
+      this.max_speed = 100
+    },
+    setSpeed(speed: number): void {
+      if (speed > this.max_speed) {
+        this.max_speed = speed
+      }
+      this.speed = speed
+    },
+    updateState(state: State): void {
+      this.state = state
+      switch (this.state) {
+        case State.DownloadSpeed:
+          this.clearSpeed()
+          this.checkDownloadSpeed()
+          break
+        case State.UploadSpeed:
+          this.clearSpeed()
+          this.checkUploadSpeed()
+          break
+        default:
+          break
+      }
+    },
+    async checkUploadSpeed(): Promise<void> {
+      const one_hundred_mega_bytes = 100 * 2 ** 20
+      const start_time = new Date().getTime()
+
+      await back_axios({
+        method: 'post',
+        url: '/network-test/post_file',
+        timeout: 20000,
+        data: new ArrayBuffer(one_hundred_mega_bytes),
+        onUploadProgress: (progress_event) => {
+          const seconds = (new Date().getTime() - start_time) / 1000
+          const speed_MB = progress_event.loaded / seconds / 2 ** 20
+          this.upload_speed = speed_MB
+          const percentage = 100 * (progress_event.loaded / progress_event.total)
+          console.debug(
+            `network-test: Upload: ${speed_MB.toFixed(2)}MB/s ${percentage.toFixed(2)}% ${seconds.toFixed(2)}s`,
+          )
+          this.setSpeed(speed_MB)
+        },
+      }).catch((error) => {
+        const message = `Failed to do speed test: ${error.message}`
+        notifications.pushError({ service: network_speed_test_service, type: 'NETWORK_SPEED_TEST_UPLOAD', message })
+        console.error(message)
+      })
+
+      this.updateState(State.Done)
+    },
+    async checkDownloadSpeed(): Promise<void> {
+      const one_hundred_mega_bytes = 100 * 2 ** 20
+
+      const start_time = new Date().getTime()
+      await back_axios({
+        method: 'get',
+        url: '/network-test/get_file',
+        timeout: 20000,
+        data: {
+          size: one_hundred_mega_bytes,
+          avoid_cache: new Date().getTime(),
+        },
+        onDownloadProgress: (progress_event) => {
+          const seconds = (new Date().getTime() - start_time) / 1000
```

Comment:
`date-fns` (which we already use) has some cool time helpers:

```javascript
start_date = new Date()
const seconds = differenceInSeconds(new Date(), start_date)
```

Reply:
Nice!